### PR TITLE
fix: upgrade to latest AWS SDK

### PR DESCRIPTION
**Pull request by sam-goodwin**
Still doesn't compile ... types need work.

I am considering:
1. dropping aws-sdk support
2. dropping attribute values (document client only) since it would be faster and easier to support

Errors:
```ts
 FAIL  test/v3-document.test.ts
  ● Test suite failed to run

    test/v3-document.test.ts:53:13 - error TS2339: Property 'key' does not exist on type 'never'.

    53   get.Item?.key;
                   ~~~

 FAIL  test/v3.test.ts
  ● Test suite failed to run

    test/v3.test.ts:43:13 - error TS2339: Property 'key' does not exist on type 'never'.

    43   get.Item?.key;
                   ~~~
    test/v3.test.ts:162:28 - error TS2339: Property 'key' does not exist on type 'never'.

    162   returnAllNew.Attributes?.key?.S;
                                   ~~~
    test/v3.test.ts:179:28 - error TS2339: Property 'key' does not exist on type 'never'.

    179   returnAllOld.Attributes?.key?.S;
                                   ~~~
    test/v3.test.ts:196:32 - error TS2339: Property 'key' does not exist on type 'never'.

    196   returnUpdatedNew.Attributes?.key?.S;
                                       ~~~
    test/v3.test.ts:213:32 - error TS2339: Property 'key' does not exist on type 'never'.

    213   returnUpdatedOld.Attributes?.key?.S;
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,31 +23,31 @@
...
-    "@aws-sdk/client-dynamodb": "^3.100.0",
-    "@aws-sdk/lib-dynamodb": "^3.100.0",
-    "@aws-sdk/smithy-client": "^3.99.0",
-    "@aws-sdk/types": "^3.78.0",
-    "@aws-sdk/util-dynamodb": "^3.100.0",
-    "@types/aws-lambda": "^8.10.97",
-    "aws-sdk": "^2.1145.0"
+    "@aws-sdk/client-dynamodb": "^3.429.0",
+    "@aws-sdk/lib-dynamodb": "^3.429.0",
+    "@smithy/smithy-client": "^3.370.0",
```
Comment (sizuhiko) [inline package.json:46]: My closed PR diff are followings.

```
    "@aws-sdk/client-dynamodb": "^3.398.0",
    "@aws-sdk/lib-dynamodb": "^3.398.0",
    "@aws-sdk/types": "^3.398.0",
    "@aws-sdk/util-dynamodb": "^3.398.0",
    "@smithy/smithy-client": "^2.0.5",
```

`@smithy/smithy-client` not have major version 3.

Comment (sizuhiko) [reply package.json:46]: Now, latest version is `2.1.11`.

Comment (sam-goodwin) [reply package.json:46]: Did you test compile? Mine blows up

Comment (sam-goodwin) [reply package.json:46]: Smithy does look wrong - I'll fix that

Comment (sam-goodwin) [reply package.json:46]:
```ts
 FAIL  test/v3-document.test.ts
  ● Test suite failed to run

    test/v3-document.test.ts:53:13 - error TS2339: Property 'key' does not exist on type 'never'.

    53   get.Item?.key;
                   ~~~

 FAIL  test/v3.test.ts
  ● Test suite failed to run

    test/v3.test.ts:43:13 - error TS2339: Property 'key' does not exist on type 'never'.

    43   get.Item?.key;
                   ~~~
    test/v3.test.ts:162:28 - error TS2339: Property 'key' does not exist on type 'never'.

    162   returnAllNew.Attributes?.key?.S;
                                   ~~~
    test/v3.test.ts:179:28 - error TS2339: Property 'key' does not exist on type 'never'.

    179   returnAllOld.Attributes?.key?.S;
                                   ~~~
    test/v3.test.ts:196:32 - error TS2339: Property 'key' does not exist on type 'never'.

    196   returnUpdatedNew.Attributes?.key?.S;
                                       ~~~
    test/v3.test.ts:213:32 - error TS2339: Property 'key' does not exist on type 'never'.

    213   returnUpdatedOld.Attributes?.key?.S;
                                       ~~~
```

**Comment by sam-goodwin** [comment]
image

**Comment by sam-goodwin** [comment]
Ah-ha! breakthrough, the `projectionExpression` was not being `const`:

```ts
export type GetCommand<
  Item extends object,
  PartitionKey extends keyof Item,
  RangeKey extends keyof Item | undefined,
  Format extends JsonFormat
> = new <
  Key extends TableKey<Item, PartitionKey, RangeKey, Format>,
  AttributesToGet extends keyof Item | undefined,
  const ProjectionExpression extends string | undefined
>(
```

Specifically adding `const` to the type:
```ts
  const ProjectionExpression extends string | undefined
```

**Comment by sam-goodwin** [comment]
Have got it working for v3 Document type but it breaks for AttributeValue. Which is fine because I'm considering dropping support for both that and all of AWS SDK v2

**Comment by sam-goodwin** [comment]
Merging this so we at least have a working version for V3 AWS SDK.

TBH, I'm pretty tempted to re-write this package. It's way too complicated.

**Comment by sam-goodwin** [comment]
@sizuhiko this change is available at `0.2.4`. I think that as long as you use the TypeSafeDocument clients it will work.

This library needs a lot of love though. Types need to be completely re-written (simplified down) and it may be useful to add an actual client that is safe.

Perhaps integrate with my https://github.com/sam-goodwin/itty-aws. Type safe + "zero" bundle impact would be awesome

**Pull request merged**